sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs

```diff
@@ -0,0 +1,343 @@
+        }
+
+        private static bool HasImplementationForMember(INamedTypeSymbol namedSymbol, ISymbol accessedMember)
+        {
+            return namedSymbol
+                .GetMembers()
+                .Any(member =>
+                    accessedMember.Equals(
+                        accessedMember.ContainingType.FindImplementationForInterfaceMember(member)));
+        }
+
+        private class ParameterData
+        {
+            public bool ShouldReportOn { get; set; } = true;
+
+            private IParameterSymbol parameterSymbol;
+            private readonly HashSet<ITypeSymbol> usedAs = new HashSet<ITypeSymbol>();
+
+            public ParameterData(IParameterSymbol parameterSymbol)
+            {
+                this.parameterSymbol = parameterSymbol;
+            }
+
+            public void AddUsage(ITypeSymbol symbolUsedAs)
+            {
+                usedAs.Add(symbolUsedAs);
+            }
+
+            public bool MatchesIdentifier(IdentifierNameSyntax id, SemanticModel semanticModel)
+            {
+                var symbol = semanticModel.GetSymbolInfo(id).Symbol;
+                return Equals(parameterSymbol, symbol);
+            }
+
+            public Diagnostic GetRuleViolation()
+            {
+                if (ShouldReportOn)
+                {
+                    var mostGeneralType = FindMostGeneralType();
+
+                    if (!Equals(mostGeneralType, parameterSymbol.Type) &&
+                        ShouldReportOnType(mostGeneralType.GetSymbolType()))
+                    {
+                        return Diagnostic.Create(rule,
+                            parameterSymbol.Locations.First(),
+                            mostGeneralType.ToDisplayString(), parameterSymbol.Type.ToDisplayString());
+                    }
+                }
+
+                return null;
+            }
+
+            private static bool ShouldReportOnType(ITypeSymbol typeSymbol)
+            {
+                bool isExcluded =
+                       typeSymbol.Is(KnownType.System_Object) ||
+                       typeSymbol.Is(KnownType.System_ValueType) ||
+                       typeSymbol.Name.StartsWith("_", StringComparison.InvariantCulture) ||
+                       typeSymbol.Is(KnownType.System_Enum);
+
+                return !isExcluded;
+            }
+
+            private ISymbol FindMostGeneralType()
+            {
+                var mostGeneralType = parameterSymbol.Type;
+
+                if (usedAs.Count == 0)
+                {
+                    return mostGeneralType;
+                }
+
+                mostGeneralType = FindMostGeneralClass(mostGeneralType);
+                mostGeneralType = FindMostGeneralInterface(mostGeneralType);
+                return mostGeneralType;
+            }
+
+            private ITypeSymbol FindMostGeneralClass(ITypeSymbol mostGeneralType)
+            {
+                ITypeSymbol currentSymbol = mostGeneralType.BaseType;
+                do
```

Comment:
I would use `while` and not `do/while` to be sure the first `currentSymbol` is not null.